archive/android_sparse.ksy

```diff
@@ -0,0 +1,75 @@
+meta:
+  id: android_sparse
+  title: Android Sparse image
+  file-extension: img
+  tags:
+    - archive
+    - android
+  license: CC0-1.0
+  endian: le
+doc: |
+    The Android sparse format is a format to more efficiently store files
+    for for example firmware updates to save on bandwidth. Files in sparse
+    format first have to be converted back to their original format.
+
+    A tool to create images for testing can be found in the Android source code tree:
+
+    https://android.googlesource.com/platform/system/core/+/master/libsparse - img2simg.c
+
+    Note: this is not the same as the Android sparse data image format.
+doc-ref: https://android.googlesource.com/platform/system/core/+/master/libsparse/sparse_format.h
+seq:
+  - id: img_header
+    type: header
+  - id: img_header_entries
+    type: image_header_entry
+    repeat: expr
+    repeat-expr: img_header.total_chunks
+types:
+  header:
+     seq:
+       - id: magic
+         contents: [0x3a, 0xff, 0x26, 0xed]
+       - id: major_version
+         type: u2
+       - id: minor_version
+         type: u2
+       - id: file_header_size
+         type: u2
+         doc: size of file header, should be 28
+       - id: chunk_header_size
+         type: u2
+         doc: size of chunk header, should be 12
+       - id: block_size
+         type: u4
+         doc: block size in bytes, multiple of 4
+       - id: total_blocks
+         type: u4
+         doc: blocks in the original data
+       - id: total_chunks
+         type: u4
+       - id: checksum
+         type: u4
+         doc: CRC32 checksum of the original data
+  image_header_entry:
+    seq:
+      - id: chunk_type
+        type: u2
+        enum: chunk_types
+        doc: chunk type
+      - id: reserved
+        type: u2
+      - id: chunk_size
+        type: u4
+        doc: in blocks in output image
+      - id: total_size
+        type: u4
+        doc: in bytes of chunk input file including chunk header and data
+      - id: body
+        size: total_size - 12
```

Comment:
`sizeof<>`?

Reply:
Yes, this is something that I probably need to fix. I looked at the manual, but totally overlooked sizeof.

Reply:
Not sure how I should do this :-/

Reply:
> I looked at the manual, but totally overlooked sizeof.

It's not yet there, sorry for that. The only documentation right now is the issue kaitai-io/kaitai_struct#84 (but note that the "functions" `sizeof()` and `lea()` are not implemented yet). We have also written it into the 0.9 release notes, but that's it:

>   * Implement compile-time `sizeof` and `bitsizeof` operators ([#84](https://github.com/kaitai-io/kaitai_struct/issues/84))
>     * Type-based: `sizeof<u4>`, `bitsizeof<b13>`, `sizeof<user_type>`
>     * Value-based: `file_header._sizeof`, `flags._bitsizeof` (`file_header`, `flags` are fields defined in the current type)

---

> Not sure how I should do this :-/

Well, when you have all the fields that needs to be subtracted on the same level, it's gonna be clumsy - you need to subtract every one of them:

```suggestion
        size: total_size - (chunk_type._sizeof + reserved._sizeof + chunk_size._sizeof + total_size._sizeof)
```

Probably a better way would be to put all fields with fixed size into a common subtype, whose total size can then be accessed by a single `sizeof` operator.

```ksy
  image_header_entry:
    seq:
      - id: header
        type: fixed_header
      - id: body
        size: header.total_size - header._sizeof
        # size: header.total_size - sizeof<fixed_header> # if you prefer
    types:
      fixed_header:
        seq:
          - id: chunk_type
            type: u2
            enum: chunk_types
            doc: chunk type
          - id: reserved
            type: u2
          - id: chunk_size
            type: u4
            doc: in blocks in output image
          - id: total_size
            type: u4
            doc: in bytes of chunk input file including chunk header and data
```

Reply:
OK, fixed.